src/api/libcellml/analyserequationast.h

```diff
@@ -0,0 +1,354 @@
+     *    - LOGBASE: the base with respect to which the logarithm is taken;
+     *    - BVAR: the bound variable of a differential equation;
+     *  - Constants:
+     *    - TRUE: the "true" boolean;
+     *    - FALSE: the "false" boolean;
+     *    - E: Euler's number;
+     *    - PI: the πconstant;
+     *    - INF: the infinity value; or
+     *    - NAN: the not-a-number value.
+     */
+    enum class Type
+    {
+        // Assignment.
+
+        ASSIGNMENT,
+
+        // Relational and logical operators.
+
+        EQ,
+        NEQ,
+        LT,
+        LEQ,
+        GT,
+        GEQ,
+        AND,
+        OR,
+        XOR,
+        NOT,
+
+        // Arithmetic operators.
+
+        PLUS,
+        MINUS,
+        TIMES,
+        DIVIDE,
+        POWER,
+        ROOT,
+        ABS,
+        EXP,
+        LN,
+        LOG,
+        CEILING,
+        FLOOR,
+        MIN,
+        MAX,
+        REM,
+
+        // Calculus elements.
+
+        DIFF,
+
+        // Trigonometric operators.
+
+        SIN,
+        COS,
+        TAN,
+        SEC,
+        CSC,
+        COT,
+        SINH,
+        COSH,
+        TANH,
+        SECH,
+        CSCH,
+        COTH,
+        ASIN,
+        ACOS,
+        ATAN,
+        ASEC,
+        ACSC,
+        ACOT,
+        ASINH,
+        ACOSH,
+        ATANH,
+        ASECH,
+        ACSCH,
+        ACOTH,
+
+        // Piecewise statement.
+
+        PIECEWISE,
+        PIECE,
+        OTHERWISE,
+
+        // Token elements.
+
+        CI,
+        CN,
+
+        // Qualifier elements.
+
+        DEGREE,
+        LOGBASE,
+        BVAR,
+
+        // Constants.
+
+        TRUE,
+        FALSE,
+        E,
+        PI,
+        INF,
+        NAN
+    };
+
+    ~AnalyserEquationAst(); /**< Destructor */
+    AnalyserEquationAst(const AnalyserEquationAst &rhs) = delete; /**< Copy constructor */
+    AnalyserEquationAst(AnalyserEquationAst &&rhs) noexcept = delete; /**< Move constructor */
+    AnalyserEquationAst &operator=(AnalyserEquationAst rhs) = delete; /**< Assignment operator */
+
+    /**
+     * @brief Create an @c AnalyserEquationAst object.
+     *
+     * Factory method to create an @c AnalyserEquationAst. Create a blank
+     * equation AST with::
+     *
+     * @code
+     *   auto ast = libcellml::AnalyserEquationAst::create();
+     * @endcode
+     *
+     * @return A smart pointer to a @c AnalyserEquationAst object.
+     */
+    static AnalyserEquationAstPtr create() noexcept;
+
+    /**
+     * @brief Get the @c Type of this @c AnalyserEquationAst.
+     *
+     * Return the @c Type of this @c AnalyserEquationAst.
+     *
+     * @return The @c Type.
+     */
+    Type type() const;
+
+    /**
+     * @brief Set the type of this @c AnalyserEquationAst.
+     *
+     * Set the type of this @c AnalyserEquationAst.
+     *
+     * @param parent The @c AnalyserEquationAst to be set as the type of this
+     * @c AnalyserEquationAst.
+     */
+    void setType(Type type);
+
```

Comment:
should the Type be const?

Reply:
Ignore me.  I think.  Not sure if we ever pass an enum as a reference or not.

Reply:
No, we don't. Think of an `enum` as an `int`. So, we pass it directly.